File: experimental/e2ee/pairing.go
package e2ee

import (
	"crypto/ed25519"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"sync"
	"time"

	"github.com/cloudflare/circl/hpke"
)

// PairingOffer is local-only: never send Secret to the platform. Its complete
// serialized value is conveyed by the user's existing scan/paste pairing action.
type PairingOffer struct {
	Version   int    `json:"version"`
	ID        string `json:"id"`
	Machine   string `json:"machine"`
	PublicKey string `json:"public_key"`
	Secret    string `json:"secret"`
	ExpiresAt int64  `json:"expires_at"`
}

type PairingIdentity struct {
	Device      string `json:"device"`
	SigningKey  string `json:"signing_key"`
	WrappingKey string `json:"wrapping_key"`
}

// PairingInvitation is deliberately process-local; a restart cancels the offer.
// Production enrollment must atomically commit identity + consumed invitation.
type PairingInvitation struct {
	mu              sync.Mutex
	offer           PairingOffer
	private         []byte
	consumed        bool
	acceptedRequest WrappedKey
	acceptedReceipt PairingReceipt
}

func NewPairingInvitation(machine string, now time.Time) (*PairingInvitation, PairingOffer, error) {
	if !identifier.MatchString(machine) {
		return nil, PairingOffer{}, ErrInvalid
	}
	private, public, err := GenerateWrappingKey()
	if err != nil {
		return nil, PairingOffer{}, err
	}
	secret := make([]byte, 32)
	id := make([]byte, 16)
	if _, err := rand.Read(secret); err != nil {
		return nil, PairingOffer{}, err
	}
	if _, err := rand.Read(id); err != nil {
		return nil, PairingOffer{}, err
	}
	offer := PairingOffer{1, base64.RawURLEncoding.EncodeToString(id), machine, base64.RawURLEncoding.EncodeToString(public), base64.RawURLEncoding.EncodeToString(secret), now.Add(5 * time.Minute).UnixMilli()}
	return &PairingInvitation{offer: offer, private: private}, offer, nil
}

func pairingContext(offer PairingOffer, now time.Time) ([]byte, []byte, []byte, error) {
	if offer.Version != 1 || !identifier.MatchString(offer.Machine) || offer.ExpiresAt <= now.UnixMilli() || offer.ExpiresAt > now.Add(5*time.Minute).UnixMilli() {
		return nil, nil, nil, ErrInvalid
	}
	if _, err := decode(offer.ID, 16, 16); err != nil {
		return nil, nil, nil, err
	}
	public, err := decode(offer.PublicKey, 65, 65)
	if err != nil {
		return nil, nil, nil, err
	}
	secret, err := decode(offer.Secret, 32, 32)
	if err != nil {
		return nil, nil, nil, err
	}
	aad, err := json.Marshal([]any{"agentwharf.e2ee.pair.v1", offer.ID, offer.Machine, offer.PublicKey, offer.ExpiresAt})
	return aad, public, secret, err
}

func validatePairingIdentity(identity PairingIdentity) error {
	if !identifier.MatchString(identity.Device) {
		return ErrInvalid
	}
	if _, err := decode(identity.SigningKey, 32, 32); err != nil {
		return err
	}
	public, err := decode(identity.WrappingKey, 65, 65)
	if err != nil {
		return err
	}
	if _, err := hpke.KEM_P256_HKDF_SHA256.Scheme().UnmarshalBinaryPublicKey(public); err != nil {
		return ErrInvalid
	}
	return nil
}

// EncryptPairingIdentity proves possession of the out-of-band 256-bit secret.
// Account/machine authorization is separate and still required at enrollment.
type pairingProof struct {
	Identity PairingIdentity `json:"identity"`
	Proof    string          `json:"proof"`
}

func pairingProofBytes(aad []byte, identity PairingIdentity) []byte {
	canonical, _ := json.Marshal(identity)
	return append(append(append([]byte("agentwharf.e2ee.pair.proof.v1\x00"), aad...), 0), canonical...)
}

func EncryptPairingIdentity(offer PairingOffer, identity PairingIdentity, signingKey ed25519.PrivateKey, now time.Time) (WrappedKey, error) {
	if err := validatePairingIdentity(identity); err != nil {
		return WrappedKey{}, err
	}
	aad, public, secret, err := pairingContext(offer, now)
	if err != nil {
		return WrappedKey{}, err
	}
	pk, err := hpke.KEM_P256_HKDF_SHA256.Scheme().UnmarshalBinaryPublicKey(public)
	if err != nil {
		return WrappedKey{}, ErrInvalid
	}
	info := sha256.Sum256(aad)
	sender, err := wrappingSuite.NewSender(pk, info[:])
	if err != nil {
		return WrappedKey{}, ErrInvalid
	}
	enc, sealer, err := sender.SetupPSK(rand.Reader, secret, []byte(offer.ID))
	if err != nil {
		return WrappedKey{}, ErrInvalid
	}
	signingPublic, err := decode(identity.SigningKey, 32, 32)
	if err != nil || len(signingKey) != ed25519.PrivateKeySize {
		return WrappedKey{}, ErrInvalid
	}
	proof := ed25519.Sign(signingKey, pairingProofBytes(aad, identity))
	if !ed25519.Verify(signingPublic, pairingProofBytes(aad, identity), proof) {
		return WrappedKey{}, ErrInvalid
	}
	plaintext, err := json.Marshal(pairingProof{identity, base64.RawURLEncoding.EncodeToString(proof)})
	if err != nil {
		return WrappedKey{}, ErrInvalid
	}
	ciphertext, err := sealer.Seal(plaintext, aad)
	if err != nil {
		return WrappedKey{}, ErrInvalid
	}
	return WrappedKey{base64.RawURLEncoding.EncodeToString(enc), base64.RawURLEncoding.EncodeToString(ciphertext)}, nil
}

type PairingReceipt struct {
	Confirmation string `json:"confirmation"`
}

// Accept only tests decryption. Production enrollment uses AcceptCommitted.
func (inv *PairingInvitation) Accept(request WrappedKey, now time.Time) (PairingIdentity, error) {
	var accepted PairingIdentity
	_, err := inv.AcceptCommitted(request, now, func(identity PairingIdentity) error { accepted = identity; return nil })
	if err == nil && accepted.Device == "" {
		return PairingIdentity{}, ErrInvalid
	}
	return accepted, err
}

// AcceptCommitted returns a receipt only after the caller durably commits the
// identity under its independently verified local account/machine authorization.
// The commit callback must be bounded and idempotent by invitation ID. An
// ambiguous commit error fences this invitation; retry cannot enroll a new peer.
func (inv *PairingInvitation) AcceptCommitted(request WrappedKey, now time.Time, commit func(PairingIdentity) error) (PairingReceipt, error) {
	if commit == nil {
		return PairingReceipt{}, ErrInvalid
	}
	return inv.acceptEnrollment(request, now, func(identity PairingIdentity, _ PairingReceipt) error { return commit(identity) })
}

func (inv *PairingInvitation) acceptEnrollment(request WrappedKey, now time.Time, commit func(PairingIdentity, PairingReceipt) error) (PairingReceipt, error) {
	inv.mu.Lock()
	defer inv.mu.Unlock()
	if commit == nil || now.UnixMilli() >= inv.offer.ExpiresAt {
		return PairingReceipt{}, ErrInvalid
	}
	if inv.consumed {
		if request == inv.acceptedRequest && inv.acceptedReceipt.Confirmation != "" {
			return inv.acceptedReceipt, nil
		}
		return PairingReceipt{}, ErrInvalid
	}
	aad, _, secret, err := pairingContext(inv.offer, now)
	if err != nil {
		return PairingReceipt{}, err
	}
	enc, err := decode(request.Enc, 65, 65)
	if err != nil {
		return PairingReceipt{}, err
	}
	ciphertext, err := decode(request.Ciphertext, 16, 1024)
	if err != nil {
		return PairingReceipt{}, err
	}
	sk, err := hpke.KEM_P256_HKDF_SHA256.Scheme().UnmarshalBinaryPrivateKey(inv.private)
	if err != nil {
		return PairingReceipt{}, ErrInvalid
	}
	info := sha256.Sum256(aad)
	receiver, err := wrappingSuite.NewReceiver(sk, info[:])
	if err != nil {
		return PairingReceipt{}, ErrInvalid
	}
	opener, err := receiver.SetupPSK(enc, secret, []byte(inv.offer.ID))
	if err != nil {
		return PairingReceipt{}, ErrInvalid
	}
	plaintext, err := opener.Open(ciphertext, aad)
	if err != nil {
		return PairingReceipt{}, ErrInvalid
	}
	var proved pairingProof
	if json.Unmarshal(plaintext, &proved) != nil || validatePairingIdentity(proved.Identity) != nil {
		return PairingReceipt{}, ErrInvalid
	}
	identity := proved.Identity
	proof, err := decode(proved.Proof, 64, 64)
	if err != nil {
		return PairingReceipt{}, ErrInvalid
	}
	public, err := decode(identity.SigningKey, 32, 32)
	if err != nil || !ed25519.Verify(public, pairingProofBytes(aad, identity), proof) {
		return PairingReceipt{}, ErrInvalid
	}
	// Re-encoding ensures exact grammar, no duplicate/unknown members or alternate
	// encoding. The sender uses the same canonical Go/TS field order.
	canonicalProof, _ := json.Marshal(proved)
	if string(canonicalProof) != string(plaintext) {
		return PairingReceipt{}, ErrInvalid
	}
	confirmationKey := opener.Export([]byte("agentwharf.e2ee.pair.confirm.v1"), 32)
	defer clear(confirmationKey)
	mac := hmac.New(sha256.New, confirmationKey)
	canonical, _ := json.Marshal(identity)
	_, _ = mac.Write(canonical)
	receipt := PairingReceipt{Confirmation: base64.RawURLEncoding.EncodeToString(mac.Sum(nil))}
	inv.consumed = true
	clear(inv.private)
	inv.offer.Secret = ""
	if err := commit(identity, receipt); err != nil {
		return PairingReceipt{}, ErrJournal
	}
	inv.acceptedRequest = request
	inv.acceptedReceipt = receipt
	return receipt, nil
}

File: client-ts/src/experimental/pairing.ts
import { Aes256Gcm, CipherSuite, DhkemP256HkdfSha256, HkdfSha256 } from '@hpke/core'

export interface PairingOffer { version: 1; id: string; machine: string; public_key: string; secret: string; expires_at: number }
export interface PairingIdentity { device: string; signing_key: string; wrapping_key: string }
export interface PairingRequest { enc: string; ciphertext: string }
export interface PairingReceipt { confirmation: string }
export interface PendingPairing { request: PairingRequest; verifyReceipt(receipt: PairingReceipt): Promise<void> }
const suite = new CipherSuite({ kem: new DhkemP256HkdfSha256(), kdf: new HkdfSha256(), aead: new Aes256Gcm() })
function invalid(): never { throw new Error('invalid pairing request') }
function encode(data: ArrayBuffer): string { return btoa(String.fromCharCode(...new Uint8Array(data))).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '') }
function decode(value: string, size: number): Uint8Array<ArrayBuffer> {
  if (typeof value !== 'string' || value.length !== Math.ceil(size * 4 / 3) || !/^[A-Za-z0-9_-]+$/.test(value)) invalid()
  const data = Uint8Array.from(atob(value.replace(/-/g, '+').replace(/_/g, '/')), c => c.charCodeAt(0))
  if (data.length !== size || encode(data.buffer) !== value) invalid()
  return data
}
// Offer comes only from the existing local scan/paste action, never a server
// response. No caller may upload or log offer.secret or the complete offer.
export async function encryptPairingIdentity(offer: PairingOffer, identity: PairingIdentity, signingKey: CryptoKey, now = Date.now()): Promise<PairingRequest> {
  return (await preparePairing(offer, identity, signingKey, now)).request
}
export async function preparePairing(offer: PairingOffer, identity: PairingIdentity, signingKey: CryptoKey, now = Date.now()): Promise<PendingPairing> {
  try {
    if (offer.version !== 1 || !Number.isSafeInteger(offer.expires_at) || offer.expires_at <= now || offer.expires_at > now + 300000) invalid()
    if (![offer.machine, identity.device].every(value => typeof value === 'string' && /^[A-Za-z0-9_.:/-]{1,128}$/.test(value))) invalid()
    decode(offer.id, 16)
    const secret = decode(offer.secret, 32)
    const recipientPublicKey = await suite.kem.deserializePublicKey(decode(offer.public_key, 65))
    decode(identity.signing_key, 32)
    await suite.kem.deserializePublicKey(decode(identity.wrapping_key, 65))
    const aad = new TextEncoder().encode(JSON.stringify(['agentwharf.e2ee.pair.v1', offer.id, offer.machine, offer.public_key, offer.expires_at]))
    const sender = await suite.createSenderContext({ recipientPublicKey, psk: { key: secret, id: new TextEncoder().encode(offer.id) }, info: await crypto.subtle.digest('SHA-256', aad) })
    const plaintext = new TextEncoder().encode(JSON.stringify({ device: identity.device, signing_key: identity.signing_key, wrapping_key: identity.wrapping_key }))
    const domain = new TextEncoder().encode('agentwharf.e2ee.pair.proof.v1\0')
    const proofInput = new Uint8Array(domain.length + aad.length + 1 + plaintext.length)
    proofInput.set(domain)
    proofInput.set(aad, domain.length)
    proofInput.set(plaintext, domain.length + aad.length + 1)
    const proof = await crypto.subtle.sign('Ed25519', signingKey, proofInput)
    const publicKey = await crypto.subtle.importKey('raw', decode(identity.signing_key, 32), 'Ed25519', false, ['verify'])
    if (!await crypto.subtle.verify('Ed25519', publicKey, proof, proofInput)) invalid()
    const protectedIdentity = new TextEncoder().encode(JSON.stringify({ identity: { device: identity.device, signing_key: identity.signing_key, wrapping_key: identity.wrapping_key }, proof: encode(proof) }))
    const request = { enc: encode(sender.enc), ciphertext: encode(await sender.seal(protectedIdentity, aad)) }
    const confirmationBytes = await sender.export(new TextEncoder().encode('agentwharf.e2ee.pair.confirm.v1'), 32)
    const confirmationKey = await crypto.subtle.importKey('raw', confirmationBytes, { name: 'HMAC', hash: 'SHA-256' }, false, ['verify'])
    new Uint8Array(confirmationBytes).fill(0)
    return {
      request,
      async verifyReceipt(receipt: PairingReceipt): Promise<void> {
        try {
          if (Object.keys(receipt).join(',') !== 'confirmation' || !await crypto.subtle.verify('HMAC', confirmationKey, decode(receipt.confirmation, 32), plaintext)) invalid()
        } catch { invalid() }
      },
    }
  } catch { return invalid() }
}
